### ISSUE TITLE
Split gradient from descent operation for SGD

### ISSUE THREAD

**Issue by gdamaskinos** [deeplearning4j/deeplearning4j#4237]
#### Issue Description

I would like to modify the gradients (e.g., add some noise for privacy reasons) computed during the backprop of a neural net and then apply them (descent).

#### Version Information
dl4j 0.9.1

#### Contributing

I tried this fix:
```java
                if (first) { // this is for the gradient field of net to be initialised properly and avoid nullpointerexeption 
			net.fit(dataset);
			first = false;
		}
		else {
			MultiLayerNetwork temp = net.clone();
			temp.fit(dataset);
			Gradient gradients = temp.gradient();			
			// e.g., modify gradients
			net.update(gradients);
		}
```
 
but no luck. The net.update() seems to not behave as expected in this case.
The .clone() works fine as the following works:
```java
			MultiLayerNetwork temp = net.clone();
			temp.fit(dataset);
			net = temp.clone();		

```

**Comment by agibsonccc** [comment]
We have this to start with for RL: https://github.com/deeplearning4j/dl4j-examples/blob/fad2373c03d3c0fabe43358022da38e06384e44e/dl4j-examples/src/main/java/org/deeplearning4j/examples/misc/externalerrors/MultiLayerNetworkExternalErrors.java

**Comment by raver119** [comment]
Technically you can try to use GradientAccumulator interface for that.
It was written for other purposes, but it acts as proxy between calculated gradients and applied gradients.

**Comment by gdamaskinos** [comment]
@raver119 if you are referring to [this](http://apiwave.com/java/snippet/SkymindIO/deeplearning4j/24752d42244698cc434b79bdd205fbfd9256336b/deeplearning4j-scaleout.deeplearning4j-scaleout-akka.src.main.java.org.deeplearning4j.iterativereduce.akka.gradient.GradientAccumulator.java) I think it's deprecated.

@agibsonccc I was able to make it work (i.e., the network is being trained) but I have issues with the learning. The problem is that the training is very slow (similar to an erroneously small value for the learning rate).  I have tried modifying the learning rate and the batch size in the update() invoke with no luck. The learning is quite slow (comparing to just using the .fit()) or the model diverges (if lrate is too big).
Any ideas what they mean in [this example](https://github.com/deeplearning4j/dl4j-examples/blob/fad2373c03d3c0fabe43358022da38e06384e44e/dl4j-examples/src/main/java/org/deeplearning4j/examples/misc/externalerrors/MultiLayerNetworkExternalErrors.java) by: 
//Update the gradient: apply learning rate, momentum, etc
?
On a different front is there any simple way to get the actual error instead of an external error?
Current version of the code:
```java
		MultiLayerNetwork temp = net.clone();
		temp.fit(dataset);
		Gradient gradient = temp.gradient();
		//System.out.println(temp.gradient().gradient());
		gradient.gradient().muli(lrate);
        net.getUpdater().update(net, gradient, 0, 1);

        //Get a row vector gradient array, and apply it to the parameters to update the model
        INDArray updateVector = gradient.gradient();
        net.params().subi(updateVector);
```

**Comment by raver119** [comment]
No, i’m not referring to that. 

Here’s what i mean: https://github.com/deeplearning4j/deeplearning4j/blob/85d494b4533a34e1fc5f9f7158a7b012b7f88a07/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/GradientsAccumulator.java#L12 <https://github.com/deeplearning4j/deeplearning4j/blob/85d494b4533a34e1fc5f9f7158a7b012b7f88a07/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/GradientsAccumulator.java#L12>

This thing is used for gradients sharing across network.
